### DISCUSSION TITLE
How to fixed the issue - CORS Missing Allow Origin

### DISCUSSION THREAD

**Discussion by acche** [nuxt/nuxt#19667] [Questions]
### Environment

- Operating System: `Darwin`
- Node Version:     `v18.15.0`
- Nuxt Version:     `3.2.3`
- Nitro Version:    `2.2.3`
- Package Manager:  `pnpm@7.29.1`
- Builder:          `vite`
- User Config:      `-`
- Runtime Modules:  `-`
- Build Modules:    `-`

### Reproduction

pnpm dev -o

### Describe the bug

Request API fail ,

Network request with error:

`CORS Missing Allow Origin`

How to fixed it?

**Comment by NyllRE** [comment]
I have the exact same issue, I'll try to explain it:

here is a route that doesn't need anything:
/server/api/test.get.ts
```ts
export default defineEventHandler(async (event) => {
	return { data: 'WORKING' };
});
```

calling this doesn't have any issues, but now I want to use the user and I stored it in a middleware:
/server/middleware/auth.ts
```ts
/** @format */

// Nuxt will automatically read in any file in the ~/server/middleware to create server middleware for your project.
// Middleware handlers will run on every request before any other server route to add or check headers, log requests, or extend the event's request object.
// https://nuxt.com/docs/guide/directory-structure/server#server-middleware

// imports...

export default defineEventHandler(async (event) => {
	// checking if the url matches the specified endpoints
	const endpoints = [
		'/api/users',
	];

        // here we check if the request route is listed in endpoints array
	const url = event.node.req.url!;
	if (!endpoints.includes(url.replace(/^(https?:)?\/\//, ''))) {
		console.log('ingored route');
	} else {

		//=>> checking token <~ headers

		const token = event.node.req.headers['authorization']?.split(' ')[1]!;

		const decoded = decodeAccessToken(token);

		console.log(`token: ${token}`);
		console.log(`Decoded Token:`);
		console.dir(decoded);
                if (!decoded) ... // sends an error

	        let user: User | null;
	        try {
                        // gets user from the database through prisma
		        user = await getUserById(decoded.id);
		        event.context.auth = { user };
	        } catch {
		        // sends an error
	}
});
```

and now the users route will try to use the user and return necessary data:
/server/api/users.post.ts
```ts
export default defineEventHandler(async (event) => {
	const user = event.context.auth?.user;

	if (user.roleName !== 'ADMIN') {
		return 'you are not an admin';
	}
        const body = await readBody(event)


	return await userList(body.userFilters);
});

```

here is a test of fetching the test api route:
![image](https://user-images.githubusercontent.com/52743522/228073016-9ee0ea57-ec6e-440e-8c91-5d665cefb495.png)

and here is a test of fetching the users api route:
![image](https://user-images.githubusercontent.com/52743522/228073120-62eb94bd-483e-435b-b993-0e55f042f258.png)


here is my configuration and all my tests trying to fix the issue:
nuxt.config.ts:
```ts
export default defineNuxtConfig({
	modules: [
		...
		'nuxt-security',
		],
	security: {
		headers: {
			accessControlAllowMethods: '*',
			accessControlAllowHeaders: '*',
			accessControlAllowOrigin: '*',
			accessControlContentType: '*',
			accessControlAccept: '*',
			// accessControlAllowCredentials: true,
			originAgentCluster: 'cross-origin',
			crossOriginResourcePolicy: 'cross-origin',
			crossOriginOpenerPolicy: 'cross-origin',
			crossOriginEmbedderPolicy: 'unsafe-none',
			// contentSecurityPolicy: false,
		},
		corsHandler: {
			origin: '*',
			methods: '*',
			headers: '*',
		},
		// xssValidator: {
		// 	stripIgnoreTag: true,
		// 	throwError: false, // optional
		// },
		// rateLimiter: {
		// 	tokensPerInterval: 200,
		// 	interval: 'day',
		// 	fireImmediately: false,
		// 	throwError: false, // optional
		// },
		// requestSizeLimiter: {
		// 	maxRequestSizeInBytes: 1_048_576 * 10,
		// 	maxUploadFileRequestInBytes: 1_048_576 * 25,
		// 	throwError: false, // optional
		// },
	},
	vite: { // disabled those because I had an issue, the link is down
		// server: {
		// cors: {
		// 	origin: '*',
		// },
		// 	proxy: {
		// 		// with options: http://localhost:5173/api/bar -> http://jsonplaceholder.typicode.com/bar
		// 		'/api': {
		// 			target: 'https://nemati.up.railway.app/api',
		// 			changeOrigin: true,
		// 			rewrite: (path) => path.replace(/^\/api/, ''),
		// 		},
		// 	},
		// },
	},
        routeRules: {
                '/**': { // same issue mentioned in the link below
	                // cors: true, 
                },
	},
```
[the issue mentioned in the vite config](https://stackoverflow.com/questions/22343384/the-access-control-allow-origin-header-contains-multiple-values)

the issue seems to be from the middleware directly, something is wrong with the middleware and the cors I guess

**Reply by NyllRE** [reply]
here is a simple way to reproduce the issue: https://stackblitz.com/edit/nuxt-starter-ktlb9b?file=nuxt.config.ts

**Comment by acche** [comment]
The complexity config file intimidate newcomers.
But anyway, thanks @NyllRE  
I need time to figure out your answer.

**Reply by NyllRE** [reply]
just for clarification, this is not an answer, the issue is happening with me too and I don't know how to fix it myself, I wanted to provide nuxt team with more details so they can look into it. But if you figure out what specific solution could fix it I would appreciate it

**Reply by acche** [reply]
@NyllRE It seems to be an issue with Nuxt3, but nobody seems to be able to solve it. Maybe switching to a different framework would be a better option.

**Reply by NyllRE** [reply]
the issue is now on route to be solved @acche : https://github.com/unjs/nitro/issues/1095

**Reply by NyllRE** [reply]
you can follow this [discord thread](https://discord.com/channels/473401852243869706/1002541842409193563/1090491998001180693) to work around the issue temporarily

**Comment by amihhs** [comment]
```ts
// /server/middlware/cors.ts or related api
export default defineEventHandler((event) => { 
  const Origin = ['https://a.com', 'https://b.com']
  const requestOrigin = getRequestHeaders(event).origin || ''
  if (Origin.includes(requestOrigin)) {
    setResponseHeaders(event, {
      'Access-Control-Allow-Methods': 'GET,HEAD,PUT,PATCH,POST,DELETE',
      'Access-Control-Allow-Origin': requestOrigin,
      // 'Access-Control-Allow-Origin': '*',
      'Access-Control-Allow-Credentials': 'true',
      'Access-Control-Allow-Headers': '*',
      'Access-Control-Expose-Headers': '*',
    })
  }
})

```
`getRequestHeaders`  `setResponseHeaders` import from h3

**Reply by acche** [reply]
> ```ts
> ```ts
> // /server/middlware/cors.ts or related api
> export default defineEventHandler((event) => { 
>   const Origin = ['https://a.com', 'https://b.com']
>   const requestOrigin = getRequestHeaders(event).origin || ''
>   if (Origin.includes(requestOrigin)) {
>     setResponseHeaders(event, {
>       'Access-Control-Allow-Methods': 'GET,HEAD,PUT,PATCH,POST,DELETE',
>       'Access-Control-Allow-Origin': requestOrigin,
>       // 'Access-Control-Allow-Origin': '*',
>       'Access-Control-Allow-Credentials': 'true',
>       'Access-Control-Allow-Headers': '*',
>       'Access-Control-Expose-Headers': '*',
>     })
>   }
> })
> ```
> 
> 
>     
>       
>     
> 
>       
>     
> 
>     
>   
> getRequestHeaders  setResponseHeaders import from h3
> ```

@amihhs Thanks for your reply. But where should I place the code to and how to use it?
